app/design/frontend/base/default/template/boltpay/prefetch_shipping_quote.phtml

```diff
@@ -29,8 +29,42 @@ if($this->isBoltActive()) {
     <script>
         (function () {
             var xmlhttp = new XMLHttpRequest();
+
             xmlhttp.open("POST", "<?=$this->getUrl('boltpay/shipping/prefetchEstimate');?>", true);
             xmlhttp.setRequestHeader("Content-type", "application/json");
+            xmlhttp.onreadystatechange = function() {
+                if (xmlhttp.readyState == XMLHttpRequest.DONE && xmlhttp.status == 200) {
+                    var result = JSON.parse(xmlhttp.responseText);
+
+                    if (result.address_data && Object.keys(result.address_data).length > 0) {
+                        var evt = document.createEvent('HTMLEvents'),
+                            _data = result.address_data,
+                            postcode = document.getElementById('postcode'),
+                            country = document.getElementById('country'),
+                            region = document.getElementById('region');
+
+                        if (postcode) {
+                            postcode.value = _data.postcode;
+                        }
+
+                        if (country) {
+                            evt.initEvent('change', true, true);
+                            country.value = _data.country_id;
+                            country.dispatchEvent(evt);
+                        }
+
+                        if (region) {
+                            if (region.tagName.toLowerCase() === 'input') {
+                                region.value = _data.region_name;
+                            } else if (region.tagName.toLowerCase() === 'select') {
+                                evt.initEvent('change', true, true);
+                                region.value = _data.region;
+                                country.dispatchEvent(evt);
+                            }
+                        }
+                    }
+                }
+            };
             xmlhttp.send(JSON.stringify(<?=$this->getLocationEstimate()?>));
```

Comment:
Changing the page element values with the geo-location estimated ones might not be the expected functionality in many cases. Maybe a check if the values are empty before the change is needed. That seems to me like a no need feature, but I might be wrong. If I am not, I'd move the whole functionality out into the controller, and just call it from the page with a GET request. To be more precise, I'd make the controller responsible for geo-ip-location estimation call and calling the shipping and tax estimation prefetch with the result. That way you can later extend the prefetch to other addresses, not just geoloc one, but logged in customer address, and also the quote shipping address that might be already saved (user entered it on the checkout-shipping page, and then returned to cart).

Reply:
I think I agree with Daniel here. we don't probably need change in this file

Reply:
Yes, it makes sense. But current functionality changes the value in estimate form only after prefetch was done. So, if in quote exist some shipping address data the form values do not be filled. This form is always empty if the customer did not enter any address previously.

Another I want to notice that if we moved this code from js, we can catch some delay before page render because we should wait for a response from GeoIp service.

Maybe I'm aggravating and I wrong, but i guess that this async request does not freeze checkout cart page for now.

Reply:
chatted off-the-thread. I'm ok to have this if you feel strongly about this (though I'm still not sure if it's our plugin's job to do this), but we need to check if the field already contain text. Currently, it is possible that user starts entering something in the field and the plugin asynchronously fetch address from server and override user's input.